### PR TITLE
Features from params

### PR DESCRIPTION
**Pull request by Wawrzyn321**
**Description**

Use with [new enkode](pr.enkode.surge.sh).

Changes proposed in this pull request:

- Introduce features, that can be controlled based on various conditions - for now, I've added `apiGroup` one and `configMapJsonPath`. Others could be a simple single value, config map string (not object)
- Some of them the are lazy-evaluated - we wouldn't want to fire a lot of requests while starting the app, would we?

**Related issue(s)**

### CODE REVIEW

File: core/src/luigi-config/init-params/constants.js [hunk 1]
```diff
@@ -14,7 +14,6 @@ export const DEFAULT_FEATURES = Object.fromEntries(
     {
       selectors: [
         {
-          lazy: false,
```
Comment (valentinvieriu) [inline core/src/luigi-config/init-params/constants.js]: if we use isDefault, should we also use isLazy ?

File: core/src/luigi-config/features.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    data.crds.some(crd => crd.includes(selector.apiGroup)),
+  configMapJsonPath: async (selector, data) => {
+    const { namespace, name, entry, jsonPath, expectedValue } = selector;
+
+    const cmUrl = `${config.backendAddress}/api/v1/namespaces/${namespace}/configmaps/${name}`;
+    const response = await failFastFetch(cmUrl, data.authData);
+    const configMap = await response.json();
+
+    const value = jsonpath.query(
+      JSON.parse(configMap.data[entry]),
```
Comment (dariadomagala) [inline core/src/luigi-config/features.js:16]: config map doesn't have to be JSON and it fails here if it's not

Comment (dariadomagala) [reply core/src/luigi-config/features.js:16]: also, this `jsonpath.query` might fail and we don't show why exactly, we should wrap this whole thing in a try/catch

**Review by dariadomagala — CHANGES_REQUESTED** [review]

File: core/src/luigi-config/features.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import { config } from './config';
+import jsonpath from 'jsonpath';
+
+const resolvers = {
+  apiGroup: (selector, data) =>
+    data.crds.some(crd => crd.includes(selector.apiGroup)),
+  configMapJsonPath: async (selector, data) => {
+    const { namespace, name, entry, jsonPath, expectedValue } = selector;
+
+    const cmUrl = `${config.backendAddress}/api/v1/namespaces/${namespace}/configmaps/${name}`;
```
[the remark this answers is not in the document]
Comment (Wawrzyn321) [reply core/src/luigi-config/features.js:11]: Well, that's why I've introduced the "lazy" option.

I did thought about caching the CMs somehow, or (as you pointed out) getting all of them at once, but all CMs in all namespaces could be large, and what about values from secrets or other resources that could be required here?

I guess we need to discuss it more.

File: core/src/luigi-config/features.js (hunk 2 above)
Comment (Wawrzyn321) [reply core/src/luigi-config/features.js:16]: I've added a try/catch block with appropriate message, good point.

About the JSON... We assume here that the CM value will be a valid JSON, as we use JsonPath.

File: core/src/luigi-config/features.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (dariadomagala) [reply core/src/luigi-config/features.js:11]: secrets? afaik we only support CMs for now?

**Pull request closed without merging**